### PR TITLE
[wasm][debugger] Improvements in debugging in async methods

### LINKED ISSUE #77031: Blazor client debugger incorrectly reading object values as null
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

While running a Blazor Web Assembly (Hosted) project, I am running into a problem while debugging (on visual studio enterprise 2022) in the client project. Newly instantiated objects with duplicate names as other objects, in the method, appear null while debugging. If they were truly null, then null reference exceptions would be thrown when calling methods on the object. They return the correct data to the UI, so there is no way for the objects to actually be null. 

### Expected Behavior

When stepping through a project, I would expect to see the debugger highlight and display values as it should. Instead, I am seeing the object value as null, even after value assignment. Setting a default constructor with a breakpoint and debugging shows the constructor being hit, but the object value still shows null.

### Steps To Reproduce

The README file in the repository [here](https://github.com/bensmith009988/BugReportTest) includes startup instructions to launch and run the sample project. Additionally, it includes instructions to find specific debugging steps that give line-by-line instructions for breakpoints while debugging to replicate the bug.

### Exceptions (if any)

Changing the names of objects in a method to be unique, (meaning no other objects in that method or class file have the same name) makes the bug disappear.

### .NET Version

6.0.402

### Anything else?

Here is a screenshot, while mid-debugging, in a code-block that should have thrown an exception but didn't because the value is not actually null. It only appears as such.

image

### LINKED ISSUE #77481: Debuggin Blazor Webassembly project of Visual Studio 2022 does not work correctly with a Visual Basic Library .NET 6 or .NET Standard Project containing Async method.
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

Debugging function of Blazor Webassembly project (like watching variable values/evaluating in Immediate window) of Visual Studio 2022 does not work correctly with referencing a Visual Basic Library .NET 6 or .NET Standard Project containing Async method.

All debuging functions except breaking point in async methods in Visual Basic Library .NET 6 or .NET Standard Project, which is referred to C# Blazor WebAssembly project, do not work with the following message in Locals window;
```
Error processing variables: TypeError: Cannot read properties of undefined (reading 'subtype')
    at Object.isArray (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\adapter\objectPreview\index.js:58:19)
    at I.createVariableByType (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\adapter\variableStore.js:114:27)
    at Q.getChildren (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\adapter\variableStore.js:659:45)
    at process.children (node:internal/process/task_queues:96:5)
    at n._a [as getVariables] (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\adapter\variableStore.js:783:15)
    at j.result [as _onVariables] (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\adapter\debugAdapter.js:269:30)
    at t.default._onMessage (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\dap\connection.js:145:27)
```

However, debugging functions in other non-async methods in Visual Basic Library .NET 6 or .NET Standard Project work correctly.

I updated Visual Studio 2022 up to the latest version (17.3.6) and updated all nuget packages.
I also deleted obj/bin folders and did clean->rebuilding, but it did not disappeared.

I am attaching Locals Window screen shots for calling non-async method and asycn method.

The following is for none-async method;
ScreenCapture 3518

The following is for async method;
ScreenCapture 3519

The issue occurs for Visual Basic .NET Standard Library Project as well as Visual Basic .NET 6 library project.
It makes hard to develop Blazor WebAssembly apps with existing Visual Basic .NET Standard/.NET6 Library projects which utilize async keyword.

### Expected Behavior

All debugging functions - watching local variables and evaluating Immediate Window - must work with async method in  referred Visual Basic Library Projects.

### Steps To Reproduce

1. Open Visual Studio 2022

2. Create a Blazor WebAssembly Project (C# .NET 6) without ASP.NET Core Hosting

3. Add a Visual Basic Library Project (.NET 6 or .NET Standard 2.0/2.1)

4. Add project reference of Visual Basic Library Project to Blazor WebAssembly Project

5. Add a Class to Visual Basic Library Project as the followings;

```
Public Class TestService
    Public Function Test(data As String) As String
        Return data
    End Function

    Public Async Function TestAsync(data As String) As Task(Of String)
        Await Task.Delay(10)
        Return data
    End Function

End Class
```

6. Modify Program.cs of Blazor WebAssembly Project as the followings;

```
using BlazorApp1;
using Microsoft.AspNetCore.Components.Web;
using Microsoft.AspNetCore.Components.WebAssembly.Hosting;

var builder = WebAssemblyHostBuilder.CreateDefault(args);
builder.RootComponents.Add<App>("#app");
builder.RootComponents.Add<HeadOutlet>("head::after");

builder.Services.AddScoped(sp => new HttpClient { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) });

//To call VB Library Project methods
var testService = new ClassLibrary1.TestService();
var result = testService.Test("1234");
var resultAsync = await testService.TestAsync("1234");

await builder.Build().RunAsync();
```

7. Add a break point at the line of 'var result = testService.Test("1234");'

8. Start Blazor WebAssembly Project and debug into testService.Test and await testService.TestAsync.

9. You can see "data" variable value in Auto/Locals window in TestService.Test method.

10. You cannot see "data" variable value in Auto/Locals window in TestService.TestAsync method.

### Exceptions (if any)

			Error processing variables: TypeError: Cannot read properties of undefined (reading 'subtype')
    at Object.isArray (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\adapter\objectPreview\index.js:58:19)
    at I.createVariableByType (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\adapter\variableStore.js:114:27)
    at Q.getChildren (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\adapter\variableStore.js:659:45)
    at process.children (node:internal/process/task_queues:96:5)
    at n._a [as getVariables] (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\adapter\variableStore.js:783:15)
    at j.result [as _onVariables] (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\adapter\debugAdapter.js:269:30)
    at t.default._onMessage (c:\program files\microsoft visual studio\2022\enterprise\common7\ide\commonextensions\microsoft\jsdiagnostics\debugger\debugAdapter\out\src\dap\connection.js:145:27)	

### .NET Version

6.0.402

### Anything else?

Test Envrionment:
Windows 10
Visual Studio 2022
.NET 6 Blazor WebAssembly Project Template

```
C:\Program Files\Microsoft Visual Studio\2022\Enterprise>dotnet --info
.NET SDK:
 Version:   6.0.402
 Commit:    6862418796

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.402\

global.json file:
  Not found

Host:
  Version:      6.0.10
  Architecture: x64
  Commit:       5a400c212a

.NET SDKs installed:
  5.0.214 [C:\Program Files\dotnet\sdk]
  5.0.303 [C:\Program Files\dotnet\sdk]
  5.0.408 [C:\Program Files\dotnet\sdk]
  6.0.402 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.30 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

### PR DESCRIPTION
**Pull request by thaystg**
- Variables with same name in different scopes in async method (C# and VB) (Before this PR it was always showing the value of the last variable declared)
- Variables in async methods in VB language
- Method names of async methods in VB language

Fixes #77031
Fixes #77481

### CODE REVIEW

**Review by ilonatommy — APPROVED** [review]

**Review by radical — CHANGES_REQUESTED** [review]

**Comment by radical** [comment]
I pushed the changes that I suggested to https://github.com/radical/runtime/commits/aj-thays_fix_77031 .

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs [hunk 1]
```diff
@@ -1999,9 +2006,34 @@ public async Task<JArray> GetHoistedLocalVariables(int objectId, IEnumerable<JTo
...
+                    }
+                    asyncLocalsFull.Add(asyncLocal);
+                }
+                else if (fieldName.StartsWith("$"))
+                {
+                    continue;
+                }
                 else
                 {
                     asyncLocalsFull.Add(asyncLocal);
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:2039]: Should this be unconditional? It seems to be common, for the cases we don't `continue` earlier.

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs [hunk 2]
```diff
@@ -1285,6 +1287,11 @@ public async Task<string> GetPrettyMethodName(int methodId, bool isAnonymous, Ca
                             if (anonymousMethodId.LastIndexOf('_') >= 0)
                                 anonymousMethodId = klassName.Substring(klassName.LastIndexOf('_') + 1);
                         }
+                        else if (klassName.StartsWith("VB$"))
+                        {
+                            var match = regexForVBAsyncMethodName.Match(klassName);
+                            ret = ret.Insert(0, match.Groups[2].Value);
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:1293]: Check `match.Success` .

**Review by radical — APPROVED** [review]
Other than the comments, LGTM 👍

**Review by radical — APPROVED** [review]

**Pull request merged**